drivers/can/can_sender.c

```diff
@@ -0,0 +1,256 @@
+
+#include <nuttx/config.h>
+#include <nuttx/can/can_sender.h>
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: can_xmit_init
+ *
+ * Description:
+ *   Initial dev sender.
+ *
+ ****************************************************************************/
+
+void can_sender_init(FAR struct can_txcache_s *cd_sender)
+{
+#ifdef CONFIG_CAN_TXPRIORITY
+  DEBUGASSERT(CONFIG_CAN_TXFIFOSIZE > 0);
+  int i;
+
+  list_initialize(&cd_sender->tx_send);
+  list_initialize(&cd_sender->tx_free);
+  list_initialize(&cd_sender->tx_untcf);
+
+  for (i = 0; i < CONFIG_CAN_TXFIFOSIZE; i++)
+    {
+      list_add_tail(&cd_sender->tx_free, &cd_sender->tx_buffer[i].list);
+    }
+
+#else
+  cd_sender->tx_head  = 0;
+  cd_sender->tx_queue = 0;
+  cd_sender->tx_tail  = 0;
+#endif
+  return;
+}
+
+/****************************************************************************
+ * Name: can_add_sendnode
+ *
+ * Description:
+ *   Add message to sender.
+ *
+ ****************************************************************************/
+
+void can_add_sendnode(FAR struct can_txcache_s *cd_sender,
+                      FAR struct can_msg_s *msg, int msglen)
+{
+#ifdef CONFIG_CAN_TXPRIORITY
+  DEBUGASSERT(FREE_COUNT(cd_sender) > 0);
+
+  struct list_node *node;
+  struct can_msg_node_s *msg_node;
+  struct can_msg_node_s *tmp_node;
+
+  node = list_remove_head(&cd_sender->tx_free);
+  msg_node = container_of(node, struct can_msg_node_s, list);
+  memcpy(&msg_node->msg, msg, msglen);
+
+  list_for_every_entry(&cd_sender->tx_send, tmp_node,
+                        struct can_msg_node_s, list)
+    {
+      if (tmp_node->msg.cm_hdr.ch_id > msg->cm_hdr.ch_id)
+        {
+          /* Prioritize tx frame based on canid */
+
+          break;
+        }
+    }
+
+  if (&tmp_node->list == &cd_sender->tx_send)
+    {
+      /* Inserted at the end of the linked list */
+
+      list_add_tail(&cd_sender->tx_send, &msg_node->list);
+    }
+  else
+    {
+      list_add_before(&tmp_node->list, &msg_node->list);
+    }
+#else
+  memcpy(&cd_sender->tx_buffer[cd_sender->tx_tail], msg, msglen);
+
+  /* Increment the tail of the circular buffer */
+
+  cd_sender->tx_tail++;
+  if (cd_sender->tx_tail >= CONFIG_CAN_TXFIFOSIZE)
+    {
+      cd_sender->tx_tail = 0;
+    }
+#endif
+
+  return;
+}
+
+/****************************************************************************
+ * Name: can_get_msg
+ *
+ * Description:
+ *   Get send message from sender.
+ *
+ ****************************************************************************/
+
+struct can_msg_s *can_get_msg(FAR struct can_txcache_s *cd_sender)
+{
+  struct can_msg_s *msg = NULL;
+
+  /* No.. The sender should not be empty in this case */
+
+  DEBUGASSERT(!FREE_FULL(cd_sender));
```

Comment:
Better to check this after all variable as declared to get C89 compatibility

Reply:
Done